### PR TITLE
introduce clang format

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- clang format file based on https://github.com/ComputationalRadiationPhysics/contributing
  - remove clang format 12 formating commands
  - add include folder ordering based on the usage of PIConGPU/PMac
- update documentation
- format all PIConGPU/PMacc files

# Review
Please check this PR commit by commit. 
There is a commit that is coping the clang format file from our contribution repository.
A second commit is currently disabling that clang format changes the include order. 
This will break the code and will be activated in a separate PR later.
The last commit is updating the documentation.
Do *NOT* review the formating! The formating commit is committed as `Tools` user.

### CODE REVIEW

File: docs/COMMIT.md [hunk 1]
```diff
@@ -6,7 +6,23 @@ We agree on the following simple rules to make our lives easier :)
...
+  ```bash
+  find include/ share/picongpu/ share/pmacc -iname "*.def" \
+  -o -iname "*.h" -o -iname "*.cpp" -o -iname "*.cu" \
+  -o -iname "*.hpp" -o -iname "*.tpp" -o -iname "*.kernel" \
+  -o -iname "*.loader" -o -iname "*.param" -o -iname "*.unitless" \
+  | xargs clang-format-11 -i
+  ```
+ 
+To format only lines you added with `git add` you can call `git clang-format-11` before you create a commit.
+Take care that un-staged changes will not be formatted.
```
Comment (sbastrakov) [inline docs/COMMIT.md:25]: Isn't this too strict and could scare off potential contributors who are not part of the core team?

Comment (psychocoderHPC) [reply docs/COMMIT.md:25]: What do you mean, `Take care that un-staged changes will not be formatted.` is how git behaves if you try to format your changed code lines only. All that is not staged will not be formated by git.

Comment (psychocoderHPC) [reply docs/COMMIT.md:25]: If you mean the formating itself. It is possible to write a bot that formats always code provided as a PR, but I think nobody of us knows how to do it, you need resources too where the bot is running or you need to integrate it somehow into the CI.
All that will not be ready in a short or mid time frame.

Comment (sbastrakov) [reply docs/COMMIT.md:25]: > What do you mean, Take care that un-staged changes will not be formatted. is how git behaves if you try to format your changed code lines only. All code not staged will not be formated by git.

My comment was not about this sentense, but the whole highlighted part. I am just worried that this piece may seem too strict in a way that a new contributor who makes some small change (e.g. clarify comments or fix a typo) would be discouraged from doing so after reading this. E.g. a typical Windows user would not even have clang format.

A bot would of course be ideal, but I already assumed that would be too difficult to do quickly.

Comment (sbastrakov) [reply docs/COMMIT.md:25]: Maybe I also misinterpret the tone and how strict it really is. @PrometheusPi @steindev @n01r what do you think?

Comment (psychocoderHPC) [reply docs/COMMIT.md:25]: But yes everyone must format there changes else the CI will not pass see #3441 But we had this in the past too e.g. for end of line whitespaces and tabs, ...

Comment (psychocoderHPC) [reply docs/COMMIT.md:25]: I think clang format is also available on windows or?

Comment (sbastrakov) [reply docs/COMMIT.md:25]: Yes, it is. Just almost no one has it "by default"

Comment (psychocoderHPC) [reply docs/COMMIT.md:25]: using clang format in visual studio https://devblogs.microsoft.com/cppblog/clangformat-support-in-visual-studio-2017-15-7-preview-1/

Comment (psychocoderHPC) [reply docs/COMMIT.md:25]: That nobody has it can not solved by us or can only solved by a bot.

**Comment by psychocoderHPC** [comment]
can we merge this and #3441 that I can go on today with some CI refactorings.

**Review by sbastrakov — APPROVED** [review]

File: docs/COMMIT.md (hunk 1 above)
Comment (PrometheusPi) [reply docs/COMMIT.md:25]: I agree with @sbastrakov impression. The sentence `Take care that un-staged changes will not be formatted.` sound relatively strict. In the end most users should use git anyway. Thus perhaps one should write: `Please be aware that un-staged changes will not be formatted.` @sbastrakov Is this more the tone you wanted and @psychocoderHPC is this still the message you wanted to communicate?

File: include/mpiInfo/mpiInfo.cpp [hunk 2]
```diff
@@ -27,7 +27,15 @@
 #include <boost/program_options/variables_map.hpp>
 
 
-#define MPI_CHECK(cmd) {int error = cmd; if(error!=MPI_SUCCESS){printf("<%s>:%i ",__FILE__,__LINE__); throw std::runtime_error(std::string("[MPI] Error"));}}
+#define MPI_CHECK(cmd)                                                                                                \
```
Comment (PrometheusPi) [inline include/mpiInfo/mpiInfo.cpp:30]: are these white spaces required by the new style?

**Review by PrometheusPi — COMMENTED** [review]
Looks good to me. I mostly ignored the last commit since it was generated automatically. But a saw some strange white space issue which I would surprise me if this is intended.

**Comment by PrometheusPi** [comment]
Sorry did not see it was already merged. @psychocoderHPC Please still have a look at the white space issue mentioned.

File: include/mpiInfo/mpiInfo.cpp (hunk 2 above)
Comment (psychocoderHPC) [reply include/mpiInfo/mpiInfo.cpp:30]: I think so that auto formated

File: docs/COMMIT.md (hunk 1 above)
Comment (psychocoderHPC) [reply docs/COMMIT.md:25]: Ican update this sentense with the already open fix for the docu

File: include/mpiInfo/mpiInfo.cpp (hunk 2 above)
Comment (PrometheusPi) [reply include/mpiInfo/mpiInfo.cpp:30]: Okay? There are **96(!)** white spaces between `#define MPI_CHECK(cmd)` and the trailing `\`.

**Pull request merged**